File: src/back-end/FunChess/src/FunChess.Core.Chess/Pieces/Bishop.cs
using FunChess.Core.Chess.Enums;
using FunChess.Core.Chess.Extensions;
using FunChess.Core.Chess.Structs;

namespace FunChess.Core.Chess.Pieces;

internal sealed class Bishop : Piece
{
    public override bool Move(Board board, Move move)
    {
        if (!MoveIsValid(board, move, out _)) return false;

        board.InternalBoard[move.Next.Index] = board.InternalBoard[move.Previous.Index];
        board.InternalBoard[move.Previous.Index] = Cell.Empty;

        return true;
    }

    public override bool MoveIsValid(Board board, Move move, out SpecialMove specialMove)
    {
        specialMove = SpecialMove.None;
        if (move.DiffX == move.DiffY)
        {
            if (Math.Abs(move.DiffX) <= 1)
            {
                return true;
            }
            return move.DiffX < 0 ? board.InternalBoard.CheckSouthwest(move) : board.InternalBoard.CheckNortheast(move);
        }
        if (move.DiffX + move.DiffY != 0) return false;
        if (Math.Abs(move.DiffX) <= 1)
        {
            return true;
        }
        return move.DiffX < 0 ? board.InternalBoard.CheckNorthwest(move) : board.InternalBoard.CheckSoutheast(move);
    }
}

File: src/front-end/fun-chess-app/src/core/chess/pieces/Bishop.ts
import Piece from "../Piece";
import Board from "../Board";
import SpecialMove from "../enums/SpecialMove";
import Move from "../structs/Move";
import { checkNortheast, checkNorthwest, checkSoutheast, checkSouthwest } from "../utils/cells-utils";

class Bishop extends Piece {
    private constructor() {
        super(64);
    }

    public static readonly instance: Piece = new Bishop();

    public override moveIsValid(board: Board, move: Move): [boolean, SpecialMove] {
        const specialMove = SpecialMove.None;
        if (move.diffX === move.diffY) {
            if (Math.abs(move.diffX) <= 1) {
                return [true, specialMove];
            }
            return [move.diffX < 0 ? checkSouthwest(board.internalBoard, move) : checkNortheast(board.internalBoard, move), specialMove];
        }
        if (move.diffX + move.diffY !== 0) return [false, specialMove];
        if (Math.abs(move.diffX) <= 1) {
            return [true, specialMove];
        }
        return [move.diffX < 0 ? checkNorthwest(board.internalBoard, move) : checkSoutheast(board.internalBoard, move), specialMove];
    }
}

export default Bishop;